csrc/transform_replay.cpp

```diff
@@ -518,10 +518,31 @@ std::pair<TensorDomain*, size_t> TransformReplay::replayPasC(
       new_IDs.push_back(id);
     }
   }
+
+  if (producer->hasAllocation()) {
+    // Currently, we are unable to replay allocation domain, so we only support
+    // cases where no replay is needed. That is, root <= allocation <= rfactor.
+    if (producer->hasRFactor()) {
+      ir_utils::validateDomainEquivalence(
+          producer->getRootDomain(), producer->getAllocationDomain());
+      ir_utils::validateDomainEquivalence(
```

Comment:
Maybe we should rename this function so that it indicates it also validates there's an ordering between the two domains.

Reply:
I would consider this as a `TORCH_INTERNAL_ASSERT` instead of a `TORCH_CHECK`. And requiring `root <= allocation <= rfactor` is not a design, but a temporary limitation. In the future, this limitation will be removed, and the correct allocation domain will be computed.

Reply:
Should the allocation domain of a producer be changed when replayed?

Reply:
I don't think so. The allocation domain should be kept as-is for producer, instead of matching the allocation domain of the consumer. During indexing, we always replay producer as consumer, and we should have a mechanism to allow specifying the producer IDs we want to index into. The allocation domain for the producer is such a mechanism.

Reply:
So, what do you exactly mean "the correct allocation domain will be computed."?

Reply:
For example,

```C++
Producer:
root: [I1, I2, I3]
allocation: [I1, I2*I3]
leaf: [I1, I2*I3]

Consumer:
root: [b, I1, I2, I3]
leaf: [b*(I1*(I2*I3))]
```

Then the producer should be replayed as

```C++
Producer:
root: [I1, I2, I3]
allocation: [I1, I2*I3]
leaf: [I1*(I2*I3)]
```

Because the `I2*I3` before replay and after replay have different `IterDomain*` (am I correct here?), we need to find which `IterDomain*` in the replay correspond to the original `IterDomain*` of `I2*I3`, and this is what I called "compute" the allocation domain.

Reply:
I see. So, it should be equivalent but may be represented differently due to pointer changes. IIRC, we try not to replay equivalent ops, so I think `I2*I3` should have the same pointer.

Reply:
> IIRC, we try not to replay equivalent ops, so I think `I2*I3` should have the same pointer.

This is indeed great! Let me check that, and lift this restriction in a separate PR.